### DISCUSSION TITLE
Significant Delay in IOS setBackgroundMessageHandler

### DISCUSSION THREAD

**Discussion by JoeToeniskoetter** [invertase/react-native-firebase#4528] [Q&A] [answered]
I am experiencing an 8+ second delay from the time the notification  is received on the device to when the background message handler is fired. 

I've tried removing bits of code one by one to see where my problem lies but I still haven't resolved this. 

Here is my index.js
```
import {AppRegistry} from 'react-native';
//import App from './App';
import messaging from '@react-native-firebase/messaging';
import React from 'react';
import {Text} from 'react-native';

function HeadlessCheck({isHeadless}) {
  if (isHeadless) {
    return null;
  }

  return <Text>Test</Text>;
}

messaging().setBackgroundMessageHandler(async (remoteMessage) => {
  console.log('Message handled in background', remoteMessage);
});
AppRegistry.registerComponent('My App', () => HeadlessCheck);
```

Even with this heavily reduced file, I am still experiencing an 8 or so second delay. 
Any pointers on debugging this issue?

**Comment by mikehardy** [comment]
Open it in Xcode, do a release mode build on a real device, then watch it using Xcode profiling tools to see where you are spending time in native code boot. Without seeing in package.json what other items are present it's impossible to know what native code is being loaded + run despite seeing the javascript here

**Reply by JoeToeniskoetter** [reply]
Thanks Mike! I will try your suggestion and let you know my results.
In the mean time here is my package.json in case anything jumps out to you.

```
{
  "name": "My_App",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint . --ext .js,.jsx,.ts,.tsx"
  },
  "dependencies": {
    "@react-native-async-storage/async-storage": "^1.13.1",
    "@react-native-community/masked-view": "^0.1.10",
    "@react-native-community/progress-bar-android": "^1.0.3",
    "@react-native-community/progress-view": "^1.1.1",
    "@react-native-firebase/admob": "^7.6.10",
    "@react-native-firebase/app": "^8.4.7",
    "@react-native-firebase/messaging": "^7.9.2",
    "@react-navigation/bottom-tabs": "^5.9.1",
    "@react-navigation/native": "^5.7.5",
    "@react-navigation/stack": "^5.9.2",
    "react": "16.13.1",
    "react-native": "0.63.2",
    "react-native-elements": "^2.3.2",
    "react-native-gesture-handler": "^1.8.0",
    "react-native-linear-gradient": "^2.5.6",
    "react-native-pdf": "^6.2.1",
    "react-native-reanimated": "^1.13.1",
    "react-native-safe-area-context": "^3.1.8",
    "react-native-screens": "^2.11.0",
    "react-native-share": "^4.0.1",
    "react-native-splash-screen": "^3.2.0",
    "react-native-vector-icons": "^7.1.0",
    "react-native-webview": "^10.9.1",
    "rn-fetch-blob": "^0.12.0"
  },
  "devDependencies": {
    "@babel/core": "^7.8.4",
    "@babel/runtime": "^7.8.4",
    "@react-native-community/eslint-config": "^1.1.0",
    "@types/jest": "^25.2.3",
    "@types/react-native": "^0.63.2",
    "@types/react-native-share": "^3.3.0",
    "@types/react-test-renderer": "^16.9.2",
    "@typescript-eslint/eslint-plugin": "^2.27.0",
    "@typescript-eslint/parser": "^2.27.0",
    "babel-jest": "^25.1.0",
    "eslint": "^6.5.1",
    "jest": "^25.1.0",
    "metro-react-native-babel-preset": "^0.59.0",
    "react-test-renderer": "16.13.1",
    "typescript": "^3.8.3"
  },
  "jest": {
    "preset": "react-native",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ]
  }
}

```

**Reply by JoeToeniskoetter** [reply]
@mikehardy 
Not sure if this helps or not but, I am getting the following output in the debugger console 6 times before the background message handler is invoked 

```
nw_read_request_report [C5] Receive failed with error "Software caused connection abort"
```

**Reply by mikehardy** [reply]
Mmm - that is a pretty spare project, nothing jumps out. Network timeouts are suspicious though, what is being fetched on startup? Why isn't it working? No idea but the answer is probably in there...

**Reply by JoeToeniskoetter** [reply]
> Mmm - that is a pretty spare project, nothing jumps out. Network timeouts are suspicious though, what is being fetched on startup? Why isn't it working? No idea but the answer is probably in there...

Yeah this is getting pretty frustrating. The network timeouts do not happen each time a notification comes through. I am not making any network requests on startup so who knows what's going on there. If the application is in quit state instead of just in the background, a notification comes through, and you attempt to open the application right away, the application crashes.

**Reply by mikehardy** [reply]
Are you certain you are using firebase-ios-sdk 6.34.0 ? check Podfile for override and make sure it's that version. There is a related problem #4299 that may mean you want to use a specific forked version of that version even, with one patch for ios notifications until #4471  lands (should be today) - but in general all I can think of is something underlying with firebase-ios-sdk or your device+network

**Reply by JoeToeniskoetter** [reply]
@mikehardy I updated to the most recent package but this did not resolve my issue. I saw a thread about the react-native-splash-screen package potentially interfering with the onNotificationOpenedApp function. Could this package be causing my issues?

**Reply by mikehardy** [reply]
One of my strongest recommendations: Never use react-native-splash-screen. Convert to react-native-boot-splash.
Bonus points, react-native-boot-splash even has a command line utility to do the initial integration. Super easy.

That said, I doubt it will affect your problem. It will just stop you from having other problems.

**Comment by Salakar** [comment, marked as answer]
Drive by comment but the 8 second delay comes from this; https://github.com/invertase/react-native-firebase/blob/master/packages/messaging/ios/RNFBMessaging/RNFBMessaging%2BAppDelegate.m#L135

This was added in https://github.com/invertase/react-native-firebase/pull/4144 due to https://github.com/invertase/react-native-firebase/issues/4104 - added in favor of working all the time but with a delay vs not working 'some of the time'.

A better solution does need implementing along the lines of what I mentioned in https://github.com/invertase/react-native-firebase/issues/4104#issuecomment-676585931 - JS needs to tell native that it's ready for the events (it does actually already do this currently so something else there is not working as intended).

**Reply by JoeToeniskoetter** [reply]
Ah, I see. So with the current implementation there really isn't anything I can do to fix this.

**Reply by mikehardy** [reply]
That's correct - I had forgotten about that one, good on @Salakar to remember it and post that up.
It has not been a priority to find the underlying cause, and no one has seemed to mind the 8 second delay enough to post about it until now, so it has seemed "sufficient" vs other priorities. Acknowledging that's not perfect we'd love a PR that fixed it for real (there are good pointers in the comment on 4104) but it will likely need to come from a motivated person in the community. If that was you I will promise to work hard to shepherd it to merge and release though
